File: java/src/main/java/vsdk/toolkit/numericalAnalysis/quasiMonteCarlo/NiederreiterCore.java
package vsdk.toolkit.numericalAnalysis.quasiMonteCarlo;

/**
Shared Niederreiter core implementation for 31-bit and 63-bit variants.
*/
class NiederreiterCore {
    private final long[][] cj;
    private final long skip;
    private final long nBitsPow;
    private final long nBitsPow1;
    private final int dimension;
    private final int numberOfBits;
    private final long[] nied;
    private long count;

    public NiederreiterCore(
        long[][] cj,
        long skip,
        long nBitsPow,
        long nBitsPow1,
        int dimension,
        int numberOfBits) {
        this.cj = cj;
        this.skip = skip;
        this.nBitsPow = nBitsPow;
        this.nBitsPow1 = nBitsPow1;
        this.dimension = dimension;
        this.numberOfBits = numberOfBits;
        this.nied = new long[dimension];
        this.count = 0;
    }

    public long[] sample(long n) {
        long diff;

        n += skip;
        diff = n ^ count; // Contains 1s where bits in n and count differ
        int bitIndex = 0;
        while (diff != 0) {
            if ((diff & 1L) != 0) {
                nied[0] ^= cj[0][bitIndex];
                nied[1] ^= cj[1][bitIndex];
                nied[2] ^= cj[2][bitIndex];
                nied[3] ^= cj[3][bitIndex];
            }

            bitIndex++;
            diff >>>= 1;
        }

        count = n;
        return nied;
    }

    public long[] nextInRange(
        long[] idx,
        int dir,
        int nmsb,
        long msb1,
        long rmsb2) {
        long mask;
        long rmask;
        long diff;
        long c;
        long i;
        long step;
        int bitIndex;

        step = 1L << nmsb;
        mask = step - 1;
        rmask = mask << (numberOfBits - nmsb);
        msb1 &= mask;
        rmsb2 &= rmask;

        i = idx[0] + skip;
        if (dir >= 0) {
            i = (((Long.compareUnsigned(i & mask, msb1) <= 0) ? i : i + mask) & ~mask) | msb1;
        }
        else {
            i = (((Long.compareUnsigned(i & mask, msb1) >= 0) ? i : i - mask) & ~mask) | msb1;
            // Keep unsigned modular arithmetic semantics from the original implementation.
            step = -step;
        }

        c = count;
        diff = (i ^ c) & mask;
        bitIndex = 0;
        while (diff != 0) {
            if ((diff & 1L) != 0) {
                nied[1] ^= cj[1][bitIndex];
            }
            diff >>>= 1;
            bitIndex++;
        }

        do {
            diff = (i ^ c) >>> nmsb;
            bitIndex = nmsb;
            while (diff != 0) {
                if ((diff & 1L) != 0) {
                    nied[1] ^= cj[1][bitIndex];
                }
                diff >>>= 1;
                bitIndex++;
            }
            c = i;
            i += step;
            if (Long.compareUnsigned(i, nBitsPow) >= 0) {
                System.err.printf(
                    "\nOverflow in Niederreiter sequence. A %d-bit sequence is not enough???\n",
                    numberOfBits);
                return null;
            }
        }
        while ((nied[1] & rmask) != rmsb2);

        diff = c ^ count;
        bitIndex = 0;
        while (diff != 0) {
            if ((diff & 1L) != 0) {
                nied[0] ^= cj[0][bitIndex];
                nied[2] ^= cj[2][bitIndex];
                nied[3] ^= cj[3][bitIndex];
            }
            diff >>>= 1;
            bitIndex++;
        }
        count = c;

        idx[0] = count - skip;
        return nied;
    }

    public long radicalInverse(long n) {
        long inv = 0;
        long f = nBitsPow1;
        while (n != 0) {
            if ((n & 1L) != 0) {
                inv |= f;
            }
            f >>>= 1;
            n >>>= 1;
            if ((n & 1L) != 0) {
                inv |= f;
            }
            f >>>= 1;
            n >>>= 1;
            if ((n & 1L) != 0) {
                inv |= f;
            }
            f >>>= 1;
            n >>>= 1;
            if ((n & 1L) != 0) {
                inv |= f;
            }
            f >>>= 1;
            n >>>= 1;
        }
        return inv;
    }

    public void foldSample(long[] xi1, long[] xi2) {
        long u = xi1[0];
        long v = xi2[0];
        long d;
        long m;

        u = (u & ~3L) | 1L; // Clear last two bits / displace point
        v = (v & ~3L) | 1L; // So it will not lay on a cell boundary
        d = (u & v) & ~1L; // Contains 1's where folding is needed

        m = nBitsPow; // Marks most significant bits
        while (d != 0) {
            if ((d & nBitsPow1) != 0) {
                // Fold
                u = (u & m) | (~(u - 1) & ~m);
                v = (v & m) | (~(v - 1) & ~m);
            }
            m |= m >>> 1;
            d <<= 1;
        }

        xi1[0] = u;
        xi2[0] = v;
    }
}

File: typescript/base/src/main/vsdk/toolkit/numericalAnalysis/quasiMonteCarlo/NiederreiterCore.ts
export class NiederreiterCore {
  private readonly cj: bigint[][];
  private readonly skip: bigint;
  private readonly nBitsPow: bigint;
  private readonly nBitsPow1: bigint;
  private readonly dimension: number;
  private readonly numberOfBits: number;
  private readonly nied: bigint[];
  private count: bigint;

  public constructor(
    cj: bigint[][],
    skip: bigint,
    nBitsPow: bigint,
    nBitsPow1: bigint,
    dimension: number,
    numberOfBits: number
  ) {
    this.cj = cj;
    this.skip = skip;
    this.nBitsPow = nBitsPow;
    this.nBitsPow1 = nBitsPow1;
    this.dimension = dimension;
    this.numberOfBits = numberOfBits;
    this.nied = new Array<bigint>(dimension).fill(0n);
    this.count = 0n;
  }

  private mask(value: bigint): bigint {
    const wordMask = (1n << BigInt(this.numberOfBits + 1)) - 1n;
    return value & wordMask;
  }

  private compareUnsigned(a: bigint, b: bigint): number {
    const ua = this.mask(a);
    const ub = this.mask(b);
    if (ua < ub) {
      return -1;
    }
    if (ua > ub) {
      return 1;
    }
    return 0;
  }

  public sample(n: bigint): bigint[] {
    n = this.mask(n + this.skip);
    let diff = this.mask(n ^ this.count);
    let bitIndex = 0;

    while (diff !== 0n) {
      if ((diff & 1n) !== 0n) {
        this.nied[0]! ^= this.cj[0]![bitIndex]!;
        this.nied[1]! ^= this.cj[1]![bitIndex]!;
        this.nied[2]! ^= this.cj[2]![bitIndex]!;
        this.nied[3]! ^= this.cj[3]![bitIndex]!;
      }
      bitIndex++;
      diff >>= 1n;
    }

    this.count = n;
    return this.nied;
  }

  public nextInRange(
    idx: bigint[],
    dir: number,
    nmsb: number,
    msb1: bigint,
    rmsb2: bigint
  ): bigint[] | null {
    let step = 1n << BigInt(nmsb);
    const mask = step - 1n;
    const rmask = mask << BigInt(this.numberOfBits - nmsb);
    msb1 &= mask;
    rmsb2 &= rmask;

    let i = this.mask(idx[0]! + this.skip);
    if (dir >= 0) {
      const condition = this.compareUnsigned(i & mask, msb1) <= 0;
      i = this.mask(((condition ? i : this.mask(i + mask)) & this.mask(~mask)) | msb1);
    }
    else {
      const condition = this.compareUnsigned(i & mask, msb1) >= 0;
      i = this.mask(((condition ? i : this.mask(i - mask)) & this.mask(~mask)) | msb1);
      step = -step;
    }

    let c = this.count;
    let diff = this.mask((i ^ c) & mask);
    let bitIndex = 0;
    while (diff !== 0n) {
      if ((diff & 1n) !== 0n) {
        this.nied[1]! ^= this.cj[1]![bitIndex]!;
      }
      diff >>= 1n;
      bitIndex++;
    }

    do {
      diff = (i ^ c) >> BigInt(nmsb);
      bitIndex = nmsb;
      while (diff !== 0n) {
        if ((diff & 1n) !== 0n) {
          this.nied[1]! ^= this.cj[1]![bitIndex]!;
        }
        diff >>= 1n;
        bitIndex++;
      }
      c = i;
      i = this.mask(i + step);
      if (this.compareUnsigned(i, this.nBitsPow) >= 0) {
        process.stderr.write(
          `\nOverflow in Niederreiter sequence. A ${this.numberOfBits}-bit sequence is not enough???\n`
        );
        return null;
      }
    }
    while ((this.nied[1]! & rmask) !== rmsb2);

    diff = this.mask(c ^ this.count);
    bitIndex = 0;
    while (diff !== 0n) {
      if ((diff & 1n) !== 0n) {
        this.nied[0]! ^= this.cj[0]![bitIndex]!;
        this.nied[2]! ^= this.cj[2]![bitIndex]!;
        this.nied[3]! ^= this.cj[3]![bitIndex]!;
      }
      diff >>= 1n;
      bitIndex++;
    }
    this.count = c;

    idx[0] = this.mask(this.count - this.skip);
    return this.nied;
  }

  public radicalInverse(n: bigint): bigint {
    n = this.mask(n);
    let inv = 0n;
    let f = this.nBitsPow1;

    while (n !== 0n) {
      if ((n & 1n) !== 0n) {
        inv |= f;
      }
      f >>= 1n;
      n >>= 1n;
      if ((n & 1n) !== 0n) {
        inv |= f;
      }
      f >>= 1n;
      n >>= 1n;
      if ((n & 1n) !== 0n) {
        inv |= f;
      }
      f >>= 1n;
      n >>= 1n;
      if ((n & 1n) !== 0n) {
        inv |= f;
      }
      f >>= 1n;
      n >>= 1n;
    }

    return inv;
  }

  public foldSample(xi1: bigint[], xi2: bigint[]): void {
    let u = this.mask(xi1[0]!);
    let v = this.mask(xi2[0]!);

    u = (u & this.mask(~3n)) | 1n;
    v = (v & this.mask(~3n)) | 1n;
    let d = (u & v) & this.mask(~1n);

    let m = this.nBitsPow;
    while (d !== 0n) {
      if ((d & this.nBitsPow1) !== 0n) {
        u = this.mask((u & m) | (this.mask(~this.mask(u - 1n)) & this.mask(~m)));
        v = this.mask((v & m) | (this.mask(~this.mask(v - 1n)) & this.mask(~m)));
      }
      m = this.mask(m | (m >> 1n));
      d = this.mask(d << 1n);
    }

    xi1[0] = u;
    xi2[0] = v;
  }
}
